core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1480,6 +1496,38 @@ private static Map<Symbol, Symbol> computeIdentityTranslations(Assignments assig
         return outputToInput;
     }
 
+    private int countRepartitionedRemoteExchangeNodes(PlanNode root)
+    {
+        return PlanNodeSearcher
+                .searchFrom(root)
+                .where(node -> node instanceof ExchangeNode exchangeNode && exchangeNode.getScope() == REMOTE && exchangeNode.getType() != REPLICATE)
+                .recurseOnlyWhen(node -> !(node instanceof ExchangeNode exchangeNode && exchangeNode.getScope() == REMOTE))
+                .findAll()
+                .size();
+    }
+
+    private int countPartitionedConnectorSource(PlanNode root)
+    {
+        return PlanNodeSearcher
+                .searchFrom(root)
+                .where(node -> node instanceof TableScanNode tableScanNode && tableScanNode.getUseConnectorNodePartitioning().orElse(false))
+                .recurseOnlyWhen(node -> !(node instanceof ExchangeNode exchangeNode && exchangeNode.getScope() == REMOTE))
+                .findAll()
+                .size();
+    }
+
+    private Stream<CatalogHandle> collectSourceCatalogs(PlanNode root)
+    {
+        return PlanNodeSearcher
+                .searchFrom(root)
+                .where(node -> node instanceof TableScanNode)
+                .recurseOnlyWhen(node -> !(node instanceof ExchangeNode exchangeNode && exchangeNode.getScope() == REMOTE))
```

Comment:
extract `!(node instanceof ExchangeNode exchangeNode && exchangeNode.getScope() == REMOTE` as separate method (3 times used).

`not(isRemoteExchange())` or `isNotRemoteExchange()`